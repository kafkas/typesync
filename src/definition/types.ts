export type PrimitiveValueType = 'nil' | 'string' | 'boolean' | 'int' | 'timestamp';

export type LiteralValueType = {
  type: 'literal';
  value: string | number | boolean;
};

export type EnumValueType = {
  type: 'enum';
  items: {
    label: string;
    value: string | number;
  }[];
};

export type MapValueType = {
  type: 'map';
  fields: Record<string, ModelField>;
};

export type UnionValueType = ValueType[];

export type AliasValueType = string;

export type ValueType =
  | PrimitiveValueType
  | LiteralValueType
  | EnumValueType
  | MapValueType
  | UnionValueType
  | AliasValueType;

export type ModelField = {
  type: ValueType;
  optional?: boolean;
  docs?: string;
};

export type DocumentModel = {
  type: 'document';
  docs?: string;
  fields: Record<string, ModelField>;
};

export type AliasModel = {
  type: 'alias';
  docs?: string;
  value: ValueType;
};

export type Model = DocumentModel | AliasModel;

export type Definition = Record<string, Model>;
