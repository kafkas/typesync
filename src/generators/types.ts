import { PythonGeneration } from './python/index.js';
import { RulesGeneration } from './rules/index.js';
import { SwiftGeneration } from './swift/index.js';
import { TSGeneration } from './ts/index.js';

export type Generation = PythonGeneration | RulesGeneration | SwiftGeneration | TSGeneration;
