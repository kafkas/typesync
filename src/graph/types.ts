export type MermaidGraphOrientation = 'TB' | 'LR';

export type MermaidGraphLink = [nodeId: string, nodeId: string];

export interface MermaidGraph {
  orientation: MermaidGraphOrientation;
  links: MermaidGraphLink[];
}
