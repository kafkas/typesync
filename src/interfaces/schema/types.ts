export type PrimitiveValueType = {
  type: 'nil' | 'string' | 'boolean' | 'int' | 'timestamp';
};

export type LiteralValueType = {
  type: 'literal';
  value: string | number | boolean;
};

export type EnumValueType = {
  type: 'enum';
  items: {
    label: string;
    value: string | number;
  }[];
};

export type ListValueType = {
  type: 'list';
  of: ValueType;
};

export type MapValueType = {
  type: 'map';
  fields: ModelField[];
};

export type UnionValueType = {
  type: 'union';
  members: ValueType[];
};

export type AliasValueType = {
  type: 'alias';
  name: string;
};

export type ValueType =
  | PrimitiveValueType
  | LiteralValueType
  | EnumValueType
  | ListValueType
  | MapValueType
  | UnionValueType
  | AliasValueType;

export interface ModelField {
  type: ValueType;
  optional: boolean;
  name: string;
  docs: string | undefined;
}

export interface DocumentModel {
  type: 'document';
  name: string;
  docs: string | undefined;
  fields: ModelField[];
}

export interface AliasModel {
  type: 'alias';
  name: string;
  docs: string | undefined;
  value: ValueType;
}

export type Model = DocumentModel | AliasModel;

export interface Schema {
  models: Model[];
}
