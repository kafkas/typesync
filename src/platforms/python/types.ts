import { StringBuilder } from '@proficient/ds';

import { assertNever } from '../../util/assert';
import { multiply } from '../../util/multiply-str';
import { space } from '../../util/space';

function indent(indentation: number, count: number) {
  return multiply(space(indentation), count);
}

export class NoneType {
  public readonly type = 'none';

  public toString() {
    return 'None';
  }
}

export class UndefinedType {
  public readonly type = 'undefined';

  public toString() {
    return 'TypeSyncUndefined';
  }
}

export class StringType {
  public readonly type = 'string';

  public toString() {
    return 'str';
  }
}

export class BooleanType {
  public readonly type = 'bool';

  public toString() {
    return 'bool';
  }
}

export class IntType {
  public readonly type = 'int';

  public toString() {
    return 'int';
  }
}

export class DatetimeType {
  public readonly type = 'datetime';

  public toString() {
    return 'datetime.datetime';
  }
}

export type PrimitiveValueType = NoneType | UndefinedType | StringType | BooleanType | IntType | DatetimeType;

export class LiteralValueType {
  public readonly type = 'literal';

  public toString() {
    switch (typeof this.value) {
      case 'string':
        return `typing.Literal["${this.value}"]`;
      case 'number':
        // TODO: Don't allow float literals in the spec
        return `typing.Literal[${this.value}]`;
      case 'boolean':
        return `typing.Literal[${this.value ? 'True' : 'False'}]`;
      default:
        assertNever(this.value);
    }
  }

  public constructor(public readonly value: string | number | boolean) {}
}

interface EnumItem {
  label: string;
  value: string | number;
}

export class EnumValueType {
  public readonly type = 'enum';

  public toString(indentation: number) {
    const builder = new StringBuilder();
    builder.append(`class ${this.name}(enum.Enum):\n`);
    this.items.forEach(item => {
      builder.append(
        `${indent(indentation, 1)}${item.label} = ${typeof item.value === 'string' ? `"${item.value}"` : item.value}\n`
      );
    });
    builder.append(`\n`);
    return builder.toString();
  }

  public constructor(
    public readonly name: string,
    private readonly items: EnumItem[]
  ) {}
}

export class TupleValueType {
  public readonly type = 'tuple';

  public toString(indentation: number) {
    // TODO: Implement
    // const pyTypes = this.values.map(v => this.getPyTypeForValueType(v, depth)).join(', ');
    // return `tuple[${pyTypes}]`;

    return 'typing.Any';
  }

  public constructor(private readonly values: ValueType[]) {}
}

export class ListValueType {
  public readonly type = 'list';

  public toString(indentation: number) {
    // TODO: Implement
    // const pyType = this.getPyTypeForValueType(type.of, depth);
    // return `typing.List[${pyType}]`;

    return 'typing.Any';
  }

  public constructor(private readonly of: ValueType) {}
}

export class MapValueType {
  public readonly type = 'map';

  public toString(indentation: number) {
    // TODO: Implement

    return 'typing.Any';
  }

  public constructor(private readonly fields: ModelField[]) {}
}

export class UnionValueType {
  public readonly type = 'union';

  public toString(indentation: number) {
    // TODO: Implement

    // const pyTypes: string[] = type.members.map(memberValueType => {
    //   return this.getPyTypeForValueType(memberValueType, depth);
    // });
    // return `typing.Union[${pyTypes.join(', ')}]`;

    return 'typing.Any';
  }

  public constructor(private readonly members: ValueType[]) {}
}

export class AliasValueType {
  public readonly type = 'alias';

  public toString() {
    return this.name;
  }

  public constructor(public readonly name: string) {}
}

export type ValueType =
  | PrimitiveValueType
  | LiteralValueType
  | EnumValueType
  | TupleValueType
  | ListValueType
  | MapValueType
  | UnionValueType
  | AliasValueType;

export interface ModelField {
  type: ValueType;
  optional: boolean;
  name: string;
  docs: string | undefined;
}
