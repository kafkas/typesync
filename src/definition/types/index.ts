export type * as types from './_types';
