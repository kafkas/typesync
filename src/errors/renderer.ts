export class RulesRendererError extends Error {
  public constructor(message: string) {
    super(`The Rules renderer encountered an error. ${message}`);
  }
}

export class InvalidOutputFileError extends Error {
  public constructor(pathToOutputFile: string, message: string) {
    super(`The output file '${pathToOutputFile}' is not valid. ${message}`);
  }
}

export class MissingRulesOutputFileError extends Error {
  public constructor(pathToOutputFile: string) {
    super(
      `The output file '${pathToOutputFile}' does not exist. An existing output file is required for Security Rules generation since it contains the markers between which the generated code is inserted.`
    );
  }
}

export class MissingStartMarkerError extends InvalidOutputFileError {
  public constructor(pathToOutputFile: string, startMarker: string) {
    super(pathToOutputFile, `The start marker '${startMarker}' is missing from the output file.`);
  }
}

export class MissingEndMarkerError extends InvalidOutputFileError {
  public constructor(pathToOutputFile: string, endMarker: string) {
    super(pathToOutputFile, `The end marker '${endMarker}' is missing from the output file.`);
  }
}

export class MisplacedStartMarkerError extends InvalidOutputFileError {
  public constructor(pathToOutputFile: string, startMarker: string, endMarker: string) {
    super(
      pathToOutputFile,
      `The start marker '${startMarker}' must be placed before the end marker '${endMarker}' in the output file.`
    );
  }
}
