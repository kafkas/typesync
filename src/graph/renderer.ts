import { StringBuilder } from '@proficient/ds';

import { MermaidGraph } from './types.js';

const TAB = '    ';

export function renderMermaidGraph(graph: MermaidGraph): string {
  const b = new StringBuilder();
  b.append(`graph ${graph.orientation}` + `\n`);
  graph.links.forEach(link => {
    const [leftNodeId, rightNodeId] = link;
    b.append(TAB + leftNodeId + ' --> ' + rightNodeId + `\n`);
  });
  return b.toString();
}
