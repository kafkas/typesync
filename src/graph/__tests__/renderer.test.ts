import { renderMermaidGraph } from '../renderer.js';
import { MermaidGraph } from '../types.js';

describe('graph', () => {
  it(`renders graph correctly`, () => {
    const graph: MermaidGraph = {
      orientation: 'LR',
      links: [
        ['books', 'generic_bookId[bookId]'],
        ['generic_bookId[bookId]', 'reviews'],
        ['generic_bookId[bookId]', 'chapters'],
        ['generic_bookId[bookId]', 'translations'],
      ],
    };
    const renderedGraph = renderMermaidGraph(graph);
    expect(renderedGraph).toMatchSnapshot();
  });
});
