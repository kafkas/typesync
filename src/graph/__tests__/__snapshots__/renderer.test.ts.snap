// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`graph renders graph correctly 1`] = `
"graph LR
    books --> generic_bookId[bookId]
    generic_bookId[bookId] --> reviews
    generic_bookId[bookId] --> chapters
    generic_bookId[bookId] --> translations
"
`;
